File: sdks/go/host.go
package wafer

import (
	"encoding/json"
	"unsafe"
)

// ---------------------------------------------------------------------------
// Host imports — declared with //go:wasmimport so TinyGo emits the correct
// WASM import section entries. These are linked by the WAFER wasmi runtime.
// ---------------------------------------------------------------------------

//go:wasmimport wafer __wafer_host_is_cancelled
func waferHostIsCancelled() int32

//go:wasmimport wafer __wafer_host_log
func waferHostLog(levelPtr, levelLen, msgPtr, msgLen int32)

//go:wasmimport wafer __wafer_host_call_block
func waferHostCallBlock(namePtr, nameLen, msgPtr, msgLen int32) int64

// ---------------------------------------------------------------------------
// Public wrappers
// ---------------------------------------------------------------------------

// IsCancelled reports whether the current execution has been cancelled by the
// runtime (e.g. client disconnect or deadline exceeded).
func IsCancelled() bool {
	return waferHostIsCancelled() != 0
}

// Log emits a log message at the given level. Conventional level strings are
// "trace", "debug", "info", "warn", and "error".
func Log(level, msg string) {
	if level == "" || msg == "" {
		return
	}
	levelBytes := []byte(level)
	msgBytes := []byte(msg)
	waferHostLog(
		int32(uintptr(unsafe.Pointer(&levelBytes[0]))), int32(len(levelBytes)),
		int32(uintptr(unsafe.Pointer(&msgBytes[0]))), int32(len(msgBytes)),
	)
}

// CallBlock calls another block by name and returns its BlockResult.
// msg is JSON-marshalled before the call; the result is JSON-unmarshalled.
// The packed (ptr << 32 | len) return value from the host points to a
// JSON-encoded BlockResult allocated in guest memory via __wafer_alloc.
func CallBlock(name string, msg *Message) BlockResult {
	if name == "" {
		panic("wafer: CallBlock called with empty block name")
	}
	nameBytes := []byte(name)
	msgData, err := json.Marshal(msg)
	if err != nil {
		panic("wafer: CallBlock marshal message: " + err.Error())
	}
	if len(msgData) == 0 {
		panic("wafer: CallBlock: marshalled message is empty")
	}

	packed := waferHostCallBlock(
		int32(uintptr(unsafe.Pointer(&nameBytes[0]))), int32(len(nameBytes)),
		int32(uintptr(unsafe.Pointer(&msgData[0]))), int32(len(msgData)),
	)

	ptr, length := unpackPtrLen(packed)
	resultData := ptrToBytes(int32(ptr), int32(length))

	// Copy before unmarshal — the slice aliases memory owned by the guest
	// allocator that the runtime may reuse.
	buf := make([]byte, len(resultData))
	copy(buf, resultData)

	var result BlockResult
	if err := json.Unmarshal(buf, &result); err != nil {
		panic("wafer: CallBlock unmarshal BlockResult: " + err.Error())
	}
	return result
}

File: sdks/js/src/host.ts
// Host function stubs for the WAFER JS/TS SDK.
//
// These functions are placeholders that throw at runtime in a plain Node/browser
// environment. When the WASM compilation pipeline is available, they will be
// replaced by imports linked by the WAFER wasmi runtime (similar to the Go SDK's
// //go:wasmimport directives).
//
// Until then, this file provides correct type signatures so block authors can
// write and type-check their code without a working WASM toolchain.

import type { BlockResult, Message } from "./types.js";

/**
 * Call another block by name and return its BlockResult.
 * Only available inside a compiled WASM block linked to the WAFER runtime.
 */
export function callBlock(_name: string, _msg: Message): BlockResult {
  throw new Error(
    "callBlock is only available inside a compiled WASM block linked to the WAFER runtime.",
  );
}

/**
 * Emit a log message at the given level.
 * Conventional level strings: "trace", "debug", "info", "warn", "error".
 * Only available inside a compiled WASM block.
 */
export function log(_level: string, _msg: string): void {
  throw new Error(
    "log is only available inside a compiled WASM block linked to the WAFER runtime.",
  );
}

/**
 * Report whether the current execution has been cancelled by the runtime
 * (e.g. client disconnect or deadline exceeded).
 * Only available inside a compiled WASM block.
 */
export function isCancelled(): boolean {
  throw new Error(
    "isCancelled is only available inside a compiled WASM block linked to the WAFER runtime.",
  );
}
